### PR TITLE
Notifications: Add support for webhooks

### PR DESCRIPTION
**Pull request by kiblik**
We would like to do an implementation similar to this: https://docs.gitlab.com/ee/user/project/integrations/webhooks.html

Events will be published via notifications

The whole functionality is described in the documentation section. This is an image which is rendered by `mermaid`

image

### CODE REVIEW

**Comment by Maffooch** [comment]
@kiblik this is looking really interesting so far. The only critique I have thus far is that it appears there is only support for one web hook at a time. I think it would be good to allow for an arbitrary amount of web hooks. Sorta like how the Jira integration works where you can connect an arbitrary amount of Jira servers and push to them according to how they are configured.

**Comment by kiblik** [comment]
@Maffooch, thank you for the feedback. I can add management of multiple webhook endpoints.

Btw, what do you think about User notifications? Do they make sense to you? They are disabled for MS Teams.

Or the combination of your idea and my question: One endpoint System notification endpoint and each user will be able to define his/her/they own endpoint.

**Comment by Maffooch** [comment]
I would think adding a web hook url would require a higher permission (I think maintainer would be sufficient) so adding user levels would be a little redundant.

**Comment by kiblik** [comment]
> I would think adding a web hook url would require a higher permission (I think maintainer would be sufficient) so adding user levels would be a little redundant.

@Maffooch, we see a possible use case when the user should be able to define his own endpoint.
What about this: 
- add `user = ForeignKey(Dojo_User, null=True, blank=True)` to `Webhook_Endpoints`
- regular user is able to create an endpoint assigned only to himself
- high-privileged user is able to assign an endpoint to anybody and also to nobody
- if it will be assigned to the user, user notifications will be processed/send (the user will also define from which products he/she/they would like to receive notifications)
- if it will not be assigned to the user, system notifications will be processed/send

**Comment by Maffooch** [comment]
I think that sounds reasonable

**Comment by Maffooch** [comment]
Hey @kiblik how are things going on this? Can I be of any help on this PR (or any of your others)?

**Comment by kiblik** [comment]
> Hey @kiblik how are things going on this? Can I be of any help on this PR (or any of your others)?

So, my priorities were on the other tasks recently but I had planned to come back to this functionality during the next week.

**Comment by nibiwodong** [comment]
Add support for webhooks is really a useful function. Looking forward to releasing this feature

**Comment by rkg-mm** [comment]
Is it possible to say when this will be available? We would like to have a 2-way sync for Dependency-Track tool, which so far only supports from DT into DD. The Webhook would be a great way to implement this without touching DD code.
https://github.com/DependencyTrack/dependency-track/issues/1926

**Comment by Maffooch** [comment]
Hey @kiblik i'v been seeing you pick this one up from time to time. How are things going on it?

**Comment by kiblik** [comment]
> Hey @kiblik i'v been seeing you pick this one up from time to time. How are things going on it?

Hi @Maffooch. Well, every time, when I wanted to continue with implementation, some other priority came in the way.
However, I was able to check it again this week and that is the reason why I asked also related [questions](https://owasp.slack.com/archives/C014H3ZV9U6/p1696975356470879).
Now, I should be able to come back and focus on it during the following 2 or 3 weeks.

**Comment by tobmay** [comment]
Is there any progress on this feature? Would love to use these generic event webhooks. 

My use case: update finding status in sources such as AWS SecHub (but also those where not integrations are available).
This is notably important for findings that are not fixed (false-positive or risk accepted), when using only DefectDojo for finding review / management...

**Comment by kiblik** [comment]
> Is there any progress on this feature? Would love to use these generic event webhooks.
> 
> My use case: update finding status in sources such as AWS SecHub (but also those where not integrations are available). This is notably important for findings that are not fixed (false-positive or risk accepted), when using only DefectDojo for finding review / management...

Hi @tobmay,
Thank you for your question. Last few days I finally found space to finish this. The core implementation is done and just need to finish the documentation. When it is ready, I will mark PR as ready and the community might start with a review. It is hard to estimate the exact date of the release but I hope it will be part of 2.38.0 (beginning of September). But no promises.

**Comment by kiblik** [comment]
@Maffooch, @mtesauro, @cneill
The core implementation of webhooks is finally (after more than a year of postponing) ready for review.
I would like to ask for feedback.

**Comment by mtesauro** [comment]
@kiblik I'm about to have a week holiday so it maybe next, next week before I can look more closely at this :+1:

**Comment by kiblik** [comment]
> @kiblik I'm about to have a week holiday so it maybe next, next week before I can look more closely at this :+1: 

No problem. I'm going to have 2 weeks of holiday. I wanted to finish it before I will leave 😀

**Comment by mtesauro** [comment]
@kiblik I could have sworn I did this already today but I don't see it above.

We're looking at merging this promptly into dev - targeting 2.38.0  :+1:

**Comment by mtesauro** [comment]
FYI: Once AccessLint gets angry, it usually never becomes happy again. Since this isn't UI, I'm sure we're fine with that test not running.

File: dojo/templates/notifications/webhooks/subtemplates/base.tpl
Comment (Maffooch) [inline dojo/templates/notifications/webhooks/subtemplates/base.tpl:1]: I didn't want to make this comment X times across each file, so I put it here instead

For the `url` field for a given object, it would be nice to supply both the UI and API url's for a given object. This would allow for user automation to "reach out" for more data from the API more conveniently than having them build the API URL themselves

File: dojo/notifications/views.py
Comment (Maffooch) [inline dojo/notifications/views.py:164]:
```suggestion
    def preprocess_request(self, request: HttpRequest):
```
I notice that `post` func in child classes are calling `super().get(request)` and this may be a little unintuitive as it implies a `post` is handling the `get` at the same time. Consider changing the name of this function to something that is more descriptive of the behavior. I suggested `preprocess_request`, but am not wedded to it

**Review by Maffooch — CHANGES_REQUESTED** [review]
@kiblik this is a great (and super clean) start to this project 😄 I tried to be overly nitpicky on this review for the sake of raising questions/discussion

File: dojo/forms.py
Comment (kiblik) [reply dojo/forms.py:2792]: This sounds good. I will need to move `first_error` and `last_error` to `List` page as some note otherwise, the user is not able to see when an error happened. Right now, the user can see it in the `Edit` form.

File: dojo/templates/notifications/webhooks/subtemplates/base.tpl
Comment (kiblik) [reply dojo/templates/notifications/webhooks/subtemplates/base.tpl:1]: NP, Any recommendation regarding the naming convention? 
- `url` + `url_ui`
- `url` + `url_api`
- `url_ui` + `url_api`
- `url` + `url_web`
- `url_web` + `url_api`
- ...

Do you have any recommendations for rendering this kind of URL?
Right now `full_url` filter is used:
https://github.com/DefectDojo/django-DefectDojo/blob/35534897d6f57d34e1d5e1ccbdf692d823d82214/dojo/templatetags/display_tags.py#L695-L697
https://github.com/DefectDojo/django-DefectDojo/blob/35534897d6f57d34e1d5e1ccbdf692d823d82214/dojo/utils.py#L1716-L1717

File: dojo/models.py
Comment (kiblik) [reply dojo/models.py]: Done: https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/470585200fdf13deca14bf3db4a54906052ada09
I didn't "commit suggestion" because I need to update migration as well.

File: dojo/templates/dojo/view_notification_webhooks.html
Comment (kiblik) [reply dojo/templates/dojo/view_notification_webhooks.html:1]: Let's see
https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/7f64990aa153be51b5f8cb0b6c58ed4e67ec0484

File: dojo/forms.py
Comment (kiblik) [reply dojo/forms.py]: Main part done: https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/fe41ee87d7d1d332cf34417802a3396b3101ec2e
Note with a date is still needed.

File: docker-compose.override.debug.yml
Comment (kiblik) [reply docker-compose.override.debug.yml:1]: Done https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/a5ece8041998aee82bce05030280be07cc369684

File: dojo/notifications/views.py
Comment (kiblik) [reply dojo/notifications/views.py]: Done: https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/56a28e34649135dde2cb0dd6ce24644975b206a4 + https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/ba36db5af75838b0215ebf673341969f86bae80c

File: dojo/notifications/helper.py
Comment (kiblik) [reply dojo/notifications/helper.py:365]: I will need to test this but sounds good

File: dojo/notifications/helper.py
Comment (kiblik) [reply dojo/notifications/helper.py:432]: No problem, I'm happy for any recommendation, on how docs might be improved. But I hope that notes (which are filled during all changes of status) might give you quite valuable information on how (or maybe why) you get there (status code, info about timeout, ...). I expect that from any error status, you want to go only to the active state. And DD will not allow you to change status until your webhook is not returning a satisfiable response (200/201). But as I said, open to any recommendations.

File: dojo/forms.py
Comment (kiblik) [reply dojo/forms.py]: Note added: https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/5e533b44449b37d4a9737efa87da89ed49ec46e7

File: dojo/templates/dojo/view_notification_webhooks.html
Comment (kiblik) [reply dojo/templates/dojo/view_notification_webhooks.html:1]: I'm not sure it helped.

File: dojo/notifications/helper.py
Comment (Maffooch) [reply dojo/notifications/helper.py:432]: Let's play it by ear. Roll with that error status and message in place, and then if there is some level confusion, add the docs then

File: dojo/templates/notifications/webhooks/subtemplates/base.tpl
Comment (Maffooch) [reply dojo/templates/notifications/webhooks/subtemplates/base.tpl:1]: > url_ui + url_api
I think this one is most comprehensive. Thought there might be a need for a new template tag. I'ms struggling to think of a way to programmatically find the API url from the relative path of the UI without hardcoding a mapping. That may be one of the only starting points though. Luckily, there is only a few objects that would need support!

File: dojo/templates/dojo/view_notification_webhooks.html
Comment (Maffooch) [reply dojo/templates/dojo/view_notification_webhooks.html:1]: I think I had to use `aria-label` at one point, but I'm not sure if that is _actually_ what fixes it

File: dojo/templates/notifications/webhooks/subtemplates/base.tpl
Comment (kiblik) [reply dojo/templates/notifications/webhooks/subtemplates/base.tpl:1]: It looks like tags might stay. I just added `reverse` calls. The hardcoding wasn't necessary.
https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/b54ef1eae095d38df897212d2619dc790b40115c

Comment (Maffooch) [reply dojo/templates/notifications/webhooks/subtemplates/base.tpl:1]: Ooh yeah good call with the reverse!

**Review by mtesauro — APPROVED** [review]
Approved

**Review by blakeaowens — APPROVED** [review]

**Review by cneill — APPROVED** [review]
Great work on this! 😄 Excited to get it merged. I noted a few minor things on this one that I wouldn't block the PR for. We can take them as TODOs for the future so this gets into `dev` for further testing.

File: dojo/notifications/helper.py
Comment (kiblik) [reply dojo/notifications/helper.py]: @cneill, I set it to any `2xx`. https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/e4287062d6aaa733f51bce48a3038e3249b9b380
Is it acceptable?

File: dojo/templates/base.html
Comment (kiblik) [reply dojo/templates/base.html]: @cneill, I extended checker. Is it better? https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/96b1730f195aff22b097af282a20e857a81d2ae3
I tested locally - I hope it is enough for this kind of change.

File: dojo/forms.py
Comment (kiblik) [reply dojo/forms.py:2791]: Thank you for noticing. Fixed: https://github.com/DefectDojo/django-DefectDojo/pull/7311/commits/3cc64f74edbd11e7ae43d3d7768ee427dde1c9c9
I would identify it in the future when I plan to finish implementation also for non-admin users, when core functionalities will be tested by other users of Dojo.

**Comment by kiblik** [comment]
I fixed the lastest comments from @cneill. I hope this PR is (finally after 1 year, 8 months, 22 days) ready for merge 😄

**Review by cneill — APPROVED** [review]
Looks good! :tada: Thanks for taking care of those final comments

**Pull request merged**
